### PR TITLE
Log learning rate to TensorBoard when using schedules

### PR DESCRIPTION
**Pull request by lgeiger**
`keras.callbacks.LearningRateScheduler` automatically logs the learning rate to TensorBoard.

However, when using `keras.optimizers.schedules.LearningRateSchedule` for step wise learning rate scheduling there is no builtin way of observing the learning rate in TensorBoard which can be crucial for debugging and experimenting with different schedules.

This PR adds support for automatic logging of the learning rate of optimizers that use `keras.optimizers.schedules.LearningRateSchedule`.

### CODE REVIEW

**Comment by gbaned** [comment]
@lgeiger Can you please resolve conflicts? Thanks!

**Comment by lgeiger** [comment]
> @lgeiger Can you please resolve conflicts?

Sure, I will resolve the conflicts.

bf2cda0f8d77cb20df252a48699483366a0e70b7 introduced the conflicts. However the changes have been [reverted](1dc12b09d671606bfdcf59539ab660f3c157846a) and [reapplied](ba5e03c88b82ad0a7a0cefb3b50cb4d71656b636) and [reverted](b1e79c7a4f3231987469e52c8b97d44837599557) and then [reapplied](69565ec4003902794bc94e10ba5fe9469a0b3ae4) again.
@omalleyt12 Are there any major changes or reverts planned soon, or is it save to rebase on top of master?

**Comment by gbaned** [comment]
@omalleyt12 Can you please assist on above question from @lgeiger . Thanks!

**Comment by gbaned** [comment]
@lgeiger Could you please resolve the conflicts? Thanks!

**Comment by lgeiger** [comment]
> @lgeiger Could you please resolve the conflicts? Thanks!

@gbaned I resolved the conflicts.

**Comment by lgeiger** [comment]
@rchao @omalleyt12 Do you have time to take a look at this?

File: tensorflow/python/keras/callbacks.py
Comment (vandenheuvel) [inline tensorflow/python/keras/callbacks.py:2073]: Could you retrieve `_decayed_lr` instead? Then the `isinstance` call etc. is taken care of, as well as any potential decay.

Comment (lgeiger) [reply tensorflow/python/keras/callbacks.py:2073]: I didn't use `_decayed_lr` here initially since it is a private API that might not behave as expected in all cases.
E.g. using `_decayed_lr` would break usage with `tf.keras.mixed_precision.experimental.LossScaleOptimizer` since it  only wraps an optimizer and would fail when trying to access `self._hyper["learning_rate"]`. Also if users opt to implement optimizers that don't rely on a `learning_rate` hyperparameter to apply gradient updates, `_decayed_lr` would fail as well. Using `getattr(self.model.optimizer, "lr", None)` covers both cases, so I would prefer to keep it as, or am I missing something?

Comment (vandenheuvel) [reply tensorflow/python/keras/callbacks.py:2073]: Thanks for the clarification. I don't think you're missing anything. To be clear: you choose to retrieve the learning rate schedule (or constant, or callable) rather than the value that would actually be used during training (the value from `model._decayed_lr` that might have a decay applied)?

Comment (lgeiger) [reply tensorflow/python/keras/callbacks.py:2073]: Yes, I wasn't aware of the `decay` keyword argument, but it looks like it is only included for backward compatibility.

**Comment by lgeiger** [comment]
@gbaned Any updates on this? It's been 2 months since I opened the PR and a month since I last rebased. I'm happy to close the PR if changes to this code from outside contributors are not welcome but a response would be greatly appreciated.

**Comment by rchao** [comment]
@lgeiger sorry for the delay - I'd like to defer to @omalleyt12 for his familiarity with TensorBoard callback.

**Comment by lgeiger** [comment]
I rebased and resolved the conflicts, this should be ready for review again.

**Comment by rpmunoz** [comment]
Hi @lgeiger , did you find a workaround to log learning rate in Tensorboard without applying this patch? I'm using the class PiecewiseConstantDecayWithWarmup(
    tf.keras.optimizers.schedules.LearningRateSchedule) from tensorflow team and facing the same problem.

`learning_rate = learning_rate.PiecewiseConstantDecayWithWarmup(
        batch_size=BATCH_SIZE,
        epoch_size=TRAIN_EXAMPLES,
        warmup_epochs=5,
        boundaries=[30, 60, 80],
        multipliers=[1.0, 0.1, 0.01, 0.001])`

`# TODO(b/149030439) - refactor this with
# tf.keras.optimizers.schedules.PiecewiseConstantDecay + WarmupDecaySchedule.
class PiecewiseConstantDecayWithWarmup(
    tf.keras.optimizers.schedules.LearningRateSchedule):
  """Piecewise constant decay with warmup schedule."""

  def __init__(self,
               batch_size: int,
               epoch_size: int,
               warmup_epochs: int,
               boundaries: List[int],
               multipliers: List[float]):
    """Piecewise constant decay with warmup.

    Args:
      batch_size: The training batch size used in the experiment.
      epoch_size: The size of an epoch, or the number of examples in an epoch.
      warmup_epochs: The number of warmup epochs to apply.
      boundaries: The list of floats with strictly increasing entries.
      multipliers: The list of multipliers/learning rates to use for the
        piecewise portion. The length must be 1 less than that of boundaries.

    """
    super(PiecewiseConstantDecayWithWarmup, self).__init__()
    if len(boundaries) != len(multipliers) - 1:
      raise ValueError("The length of boundaries must be 1 less than the "
                       "length of multipliers")

    base_lr_batch_size = 256
    steps_per_epoch = epoch_size // batch_size

    self._rescaled_lr = BASE_LEARNING_RATE * batch_size / base_lr_batch_size
    self._step_boundaries = [float(steps_per_epoch) * x for x in boundaries]
    self._lr_values = [self._rescaled_lr * m for m in multipliers]
    self._warmup_steps = warmup_epochs * steps_per_epoch

  def __call__(self, step: int):
    """Compute learning rate at given step."""
    def warmup_lr():
      return self._rescaled_lr * (
          step / tf.cast(self._warmup_steps, tf.float32))
    def piecewise_lr():
      return tf.compat.v1.train.piecewise_constant(
          tf.cast(step, tf.float32), self._step_boundaries, self._lr_values)
    return tf.cond(step < self._warmup_steps, warmup_lr, piecewise_lr)

  def get_config(self) -> Mapping[str, Any]:
    return {
        "rescaled_lr": self._rescaled_lr,
        "step_boundaries": self._step_boundaries,
        "lr_values": self._lr_values,
        "warmup_steps": self._warmup_steps,
    }`

**Comment by lgeiger** [comment]
@rpmunoz I am currently using the following workaround with TF 2.2:
```python
class TensorBoard(tf.keras.callbacks.TensorBoard):
    def _collect_learning_rate(self, logs):
        logs = logs or {}
        lr_schedule = getattr(self.model.optimizer, "lr", None)
        if isinstance(lr_schedule, tf.keras.optimizers.schedules.LearningRateSchedule):
            logs["learning_rate"] = tf.keras.backend.get_value(
                lr_schedule(self.model.optimizer.iterations)
            )
        return logs

    def _log_metrics(self, logs, prefix, step):
        super()._log_metrics(self._collect_learning_rate(logs), prefix, step)
```

@gbaned @omalleyt12 is there anything that is blocking this from beeing reviewed?

**Comment by gbaned** [comment]
@omalleyt12 Can you please review this PR ? Thanks!

**Comment by lgeiger** [comment]
@gbaned It seems like @omalleyt12 is currently busy with other things or doesn't receive GitHub notifications, could you request a review from another member of the TensorFlow team so this PR can move forward?

It's been 4 months since the PR was opened which is not a great contributor experience (this is not my only TensorFlow PR that has been awaiting reviews for a long time which makes it even more frustrating). Just to be clear, I understand that keeping up with the amount of PRs is challenging and I am by no means trying to push towards getting the changes accepted, but I would really appreciate a review or an update on the status of this.

**Review by rchao — DISMISSED** [review]
Sorry about the wait on the review! I think it looks good to me and approving. Thanks for the PR.

**Comment by lgeiger** [comment]
> Sorry about the wait on the review! I think it looks good to me and approving. Thanks for the PR.

@rchao Thanks for the review 👍 
Looks like CI is failing due to an unrelated error, I rebased since this is probably fixed on master.

**Comment by lgeiger** [comment]
@rchao My rebase invalidated your approval, do you mind approving it again?

**Comment by rchao** [comment]
> @rchao My rebase invalidated your approval, do you mind approving it again?

Looks good. Thanks!

**Review by rchao — APPROVED** [review]

**Pull request merged**
